### ISSUE TITLE
NoReturn not propagated from method call

### ISSUE THREAD

**Issue by matangover** [microsoft/pylance-release#1369]
## Environment data

-   Language Server version: 2021.5.3 (pyright 1f57ad5c)
-   OS and version: macOS 10.15.7
-   Python version: 3.8.6

## Expected behaviour

Pylance should correctly infer that a function does not return if it calls a method that does not return.

## Actual behaviour

Pylance issues an error: `Function with declared type of "NoReturn" cannot return "None"` even though the function in question only calls a NoReturn method.

## Code Snippet / Additional information

Note that in the following example test2 is marked as error and test1 is not.

```python
from typing import NoReturn

class Class:
    def noreturn_method(self) -> NoReturn:
        raise RuntimeError()

def noreturn_function() -> NoReturn:
    raise RuntimeError()

def test1() -> NoReturn:
    noreturn_function()

def test2() -> NoReturn:
    Class().noreturn_method()

```

![image](https://user-images.githubusercontent.com/5028956/120087274-25461a00-c0ef-11eb-81ae-a53e8a32152b.png)

**Comment by erictraut** [comment]
The problem is that code flow analysis requires that we build a code flow graph and determine which paths through that graph will be traversed at runtime — and which lead to "no return" conditions. This graph is required for semantic type analysis. But sometimes the semantic types can in turn influence the code flow graph. 

To break the cycle, the type checker supports certain common expression forms when evaluating whether a call expression is a no-return type. See [here](https://github.com/microsoft/pyright/blob/0f9d308827e014f2b9b65cd4864cc0b889c53236/packages/pyright-internal/src/analyzer/typeEvaluator.ts#L16140) if you're interested in the code. It supports call expressions that consist of simple names (e.g. `foo(x, y, z)`) or member access expressions (e.g. `foo.bar(x, y, z)`), but it doesn't currently support chained call expressions like in your example above (`foo().bar(x, y, z)`).

We could add more special-case handling here if there is sufficient demand, but the current limitations have proven to work for the vast majority of cases.

You can work around this limitation by using one of the supported forms or by adding an extraneous `raise` statement in your code.

**Comment by matangover** [comment]
Thanks for the detailed explanation. I tried to use a member access expressions but it didn't work:
```python
def test5() -> NoReturn: # error still shown
    c = Class()
    c.noreturn_method()
```
Reading the comment in the code you linked to, I realized that adding a type annotation is a viable workaround:
```python
def test6() -> NoReturn: # no error shown
    c: Class = Class()
    c.noreturn_method()
```

Btw, before I simplified my code to report the bug, it looked like this (well, still simplified):
```python
from argparse import ArgumentParser

class MyArgumentParser(ArgumentParser):
    def error(self, message): # Pylance reports error here
        super().error(message)
```
The error reported is:
```
Method "error" overrides class "ArgumentParser" in an incompatible manner
  Return type mismatch: base method returns type "NoReturn", override returns type "None"
    Type "None" cannot be assigned to type "NoReturn"
```

**Comment by erictraut** [comment]
Yes, by providing a type annotation, you eliminate the need for the type checker to perform type inference, which requires code flow analysis. So that's a good workaround as well.

**Comment by matangover** [comment]
@jakebailey is this closed as fixed or wontfix?

**Comment by erictraut** [comment]
It's closed as won't fix.

**Comment by erictraut** [comment]
I've added more logic to handle this particular case, so this will be addressed in the next release of pyright.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2022.1.0, which we've just released. You can find the changelog here:  [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202210-5-january-2022)
